js/util.js

```diff
@@ -90,3 +101,41 @@ util.guessLineEndings = function(text) {
 
   return (hasCRLF ? '\r\n' : '\n');
 };
+
+/**
+ * @param {?string} opt_content Optional content.
+ * @return {SessionDescriptor}
+ * Creates a unified session that can be read from any supported editor.
+ */
+util.createUnifiedSession = function(opt_content) {
+  const textarea = document.createElement('textarea');
+  textarea.value = opt_content || '';
+
+  return {
+    codemirror: new CodeMirror.Doc(opt_content || ''),
+    textarea: textarea,
+  };
+}
+
+/**
+ * @param {SessionDescriptor} session
+ * @param {string} updated Which text source is the source of truth.
+ * @param {string} lineEndings What to use as a line ending
+ * Syncs the multiple formats of a unified session. If one format of the session
+ * such as the codemirror instance generates a change, it's registered here and
+ * copied over to the other formats (such as textarea) so all of the formats
+ * have the correct text. This means if you switch between modes you don't lose
+ * any text and the undo stack is consistent (the redo stack may
```

Comment:
what behavior/bugs do we expect from a divergent redo stack?

Reply:
Thanks for pointing this comment out, i actually think it's incorrect....
Setting the text via this way seems to remove all undo/redo stack history for textarea. Codemirror also loses it's undo and redo history but tries to reconstruct it assuming each character was a single edit.
I'll update the comment and add a todo to make it better